## Data Cleaning

In this notebook we will go through the process of cleaning the data obtained from ratebeer.com. We will look at the different columns, figure out strategies to impute missing data, drop unwanted columns and add columns as needed. 

## Import Modules

In [25]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

## Load Data

We have already acquired the data from ratebeer.com and saved the data as a .csv file already.

In [26]:
df = pd.read_csv('RateBeer.csv')

In [27]:
df.head()

Unnamed: 0  beer_id                                          beer_name  \
0           1    58057                       Närke Kaggen Stormaktsporter   
1           2     4934                              Westvleteren 12 (XII)   
2           3   231441                    Schramm’s The Heart of Darkness   
3           4   106749  B. Nektar Ken Schramm Signature Series - The H...   
4           5   140581  Cigar City Pilot Series Dragonfruit Passion Fr...   

                                    beer_description  beer_abv  \
0  Imperial Stout brewed with heather honey and a...       9.5   
1  Westvleteren has the smallest output of the Tr...      10.2   
2  The Heart of Darkness is our capstone mead. It...      14.0   
3  Meadmaker Ken Schramm crafted the Heart of Dar...      14.0   
4  Editor’s Note: This is an archived entry for t...       0.0   

   beer_styleScore  beer_overallScore  beer_averageRating  beer_ratingCount  \
0            100.0              100.0            4.489771               557   
1            100.0              100.0            4.426578              3332   
2            100.0              100.0            4.423655                77   
3            100.0              100.0            4.421873                50   
4            100.0              100.0            4.420719                46   

    beer_styleName                   brewery_name  brewery_type  \
0   Imperial Stout           Närke Kulturbryggeri  Microbrewery   
1    Abt/Quadrupel  Westvleteren Abdij St. Sixtus  Microbrewery   
2             Mead                 Schramm’s Mead       Meadery   
3             Mead              B. Nektar Meadery       Meadery   
4  Berliner Weisse             Cigar City Brewing  Microbrewery   

         brewery_street  brewery_city brewery_state brewery_country  \
0      Beväringsgatan 2        Örebro           NaN          Sweden   
1       Donkerstraat 12  Westvleteren           NaN         Belgium   
2  327 West 9 Mile Road      Ferndale      Michigan   United States   
3           1511 Jarvis      Ferndale      Michigan   United States   
4  3924 W Spruce Street         Tampa       Florida   United States   

  brewery_continent brewery_twitter                      brewery_facebook  
0               NaN             NaN                                   NaN  
1               NaN             NaN                                   NaN  
2     North America    schrammsmead  https://facebook.com/SchrammsMeadery  
3     North America         bnektar         https://facebook.com/b.nektar  
4     North America   cigarcitybeer    https://facebook.com/cigarcitybeer

In [28]:
#drop the unnamed column
df.drop(["Unnamed: 0"], axis=1, inplace=True)

In [29]:
df.head()

beer_id                                          beer_name  \
0    58057                       Närke Kaggen Stormaktsporter   
1     4934                              Westvleteren 12 (XII)   
2   231441                    Schramm’s The Heart of Darkness   
3   106749  B. Nektar Ken Schramm Signature Series - The H...   
4   140581  Cigar City Pilot Series Dragonfruit Passion Fr...   

                                    beer_description  beer_abv  \
0  Imperial Stout brewed with heather honey and a...       9.5   
1  Westvleteren has the smallest output of the Tr...      10.2   
2  The Heart of Darkness is our capstone mead. It...      14.0   
3  Meadmaker Ken Schramm crafted the Heart of Dar...      14.0   
4  Editor’s Note: This is an archived entry for t...       0.0   

   beer_styleScore  beer_overallScore  beer_averageRating  beer_ratingCount  \
0            100.0              100.0            4.489771               557   
1            100.0              100.0            4.426578              3332   
2            100.0              100.0            4.423655                77   
3            100.0              100.0            4.421873                50   
4            100.0              100.0            4.420719                46   

    beer_styleName                   brewery_name  brewery_type  \
0   Imperial Stout           Närke Kulturbryggeri  Microbrewery   
1    Abt/Quadrupel  Westvleteren Abdij St. Sixtus  Microbrewery   
2             Mead                 Schramm’s Mead       Meadery   
3             Mead              B. Nektar Meadery       Meadery   
4  Berliner Weisse             Cigar City Brewing  Microbrewery   

         brewery_street  brewery_city brewery_state brewery_country  \
0      Beväringsgatan 2        Örebro           NaN          Sweden   
1       Donkerstraat 12  Westvleteren           NaN         Belgium   
2  327 West 9 Mile Road      Ferndale      Michigan   United States   
3           1511 Jarvis      Ferndale      Michigan   United States   
4  3924 W Spruce Street         Tampa       Florida   United States   

  brewery_continent brewery_twitter                      brewery_facebook  
0               NaN             NaN                                   NaN  
1               NaN             NaN                                   NaN  
2     North America    schrammsmead  https://facebook.com/SchrammsMeadery  
3     North America         bnektar         https://facebook.com/b.nektar  
4     North America   cigarcitybeer    https://facebook.com/cigarcitybeer

In [30]:
df.shape

(220000, 18)

The data set contains 220,000 beers with 18 variables. There are 5 columns that are numerical and will be important for the analysis. Preliminary inspection of the dataset indicates that the data is fairly clean with mising values. The 'beer_description' column will be interesting to look at from a sentiment analysis perspective. However, it will not be applied for this project. Other columns such as brewery_name, brewery_street can be dropped from the data frame since they add no value for the current analysis. The brewery_country column will be useful to look at the variation in beer ratings across countries i.e. to answer the question which country produces the best beer? The brewery_continent column is too broad for an analysis and is expected not to give any insights. We will probably compare beers only from a few countries. The last two columns give the links to the beer/brewery social media accounts. These will be converted to binary columns for analysis to analyze the effect of social media on the ratings and other variables. We can drop the overall score and style score since these are percentile values that are actually based on the ratings and will not provide insights. 

In [31]:
# drop unwanted columns from the dataframe
cols = ['beer_name','beer_abv','beer_averageRating','beer_ratingCount','beer_styleName','brewery_type','brewery_city','brewery_state','brewery_country','brewery_twitter','brewery_facebook']
df = df.loc[:,cols]

In [32]:
df.head()

beer_name  beer_abv  \
0                       Närke Kaggen Stormaktsporter       9.5   
1                              Westvleteren 12 (XII)      10.2   
2                    Schramm’s The Heart of Darkness      14.0   
3  B. Nektar Ken Schramm Signature Series - The H...      14.0   
4  Cigar City Pilot Series Dragonfruit Passion Fr...       0.0   

   beer_averageRating  beer_ratingCount   beer_styleName  brewery_type  \
0            4.489771               557   Imperial Stout  Microbrewery   
1            4.426578              3332    Abt/Quadrupel  Microbrewery   
2            4.423655                77             Mead       Meadery   
3            4.421873                50             Mead       Meadery   
4            4.420719                46  Berliner Weisse  Microbrewery   

   brewery_city brewery_state brewery_country brewery_twitter  \
0        Örebro           NaN          Sweden             NaN   
1  Westvleteren           NaN         Belgium             NaN   
2      Ferndale      Michigan   United States    schrammsmead   
3      Ferndale      Michigan   United States         bnektar   
4         Tampa       Florida   United States   cigarcitybeer   

                       brewery_facebook  
0                                   NaN  
1                                   NaN  
2  https://facebook.com/SchrammsMeadery  
3         https://facebook.com/b.nektar  
4    https://facebook.com/cigarcitybeer

Now that we have only the required columns in the dataframe, we will give easily readable names to the columns.

In [33]:
df.columns = ['Name','ABV','AverageRating','RatingCount','BeerStyle','BreweryType','City','State','Country','Twitter','Facebook']
df.head()

Name   ABV  AverageRating  \
0                       Närke Kaggen Stormaktsporter   9.5       4.489771   
1                              Westvleteren 12 (XII)  10.2       4.426578   
2                    Schramm’s The Heart of Darkness  14.0       4.423655   
3  B. Nektar Ken Schramm Signature Series - The H...  14.0       4.421873   
4  Cigar City Pilot Series Dragonfruit Passion Fr...   0.0       4.420719   

   RatingCount        BeerStyle   BreweryType          City     State  \
0          557   Imperial Stout  Microbrewery        Örebro       NaN   
1         3332    Abt/Quadrupel  Microbrewery  Westvleteren       NaN   
2           77             Mead       Meadery      Ferndale  Michigan   
3           50             Mead       Meadery      Ferndale  Michigan   
4           46  Berliner Weisse  Microbrewery         Tampa   Florida   

         Country        Twitter                              Facebook  
0         Sweden            NaN                                   NaN  
1        Belgium            NaN                                   NaN  
2  United States   schrammsmead  https://facebook.com/SchrammsMeadery  
3  United States        bnektar         https://facebook.com/b.nektar  
4  United States  cigarcitybeer    https://facebook.com/cigarcitybeer

As a next step, we will look at the summary statistics for the numerical columns to understand the data.

In [34]:
df.describe()

ABV  AverageRating    RatingCount
count  220000.000000  220000.000000  220000.000000
mean        5.914893       3.090490      35.734341
std         2.597578       0.372849     127.098904
min         0.000000       1.800000       0.000000
25%         4.700000       2.878205       5.000000
50%         5.500000       3.112463      10.000000
75%         7.000000       3.340448      25.000000
max        73.000000       4.489771    5182.000000

There are beers with a ABV values of 0. This is very interesting. We need to look at these more closely. These could be errors in the data or missing values coded as 0. The mean value of ABV is about 6%. There are beers ABV as high as 73%. These are rare and maybe outliers in the data. 

The zero values in the rating count column is suspect. A beer cannot have an average rating without anybody reviewing it! We will also look at these closely.

The zero values will not be dealt with now and assume that the zero values are not missing. We will inspect these in detail when we do an exploratory data analysis. 

### Social Media

We will be using the facebook and twitter columns to analyze the influence of social media on the rating counts and average rating. We will asumme that the missing values indicate that the beer/brewery does not have a social media account. We will create the following binary columns - has_twitter and has_facebook for analysis.

In [36]:
df.Twitter.isnull().sum()

110147

In [37]:
df.Facebook.isnull().sum()

60080

There are quite a few beers with either social media accounts - 109,853 beers have twitter, 159,220 have facebook. Looks like brewries prefer facbook over twitter. This can be an interesting analysis to see which social media account has a greater influence on the average rating. Let's create the binary columns mentioned above.

In [52]:
import numpy as np
df['has_twitter'] = np.where(pd.isnull(df.Twitter), 0, 1)

In [53]:
df['has_facebook'] = np.where(pd.isnull(df.Facebook), 0, 1)

Lets quickly check the number of beers with twitter and facebook.

In [54]:
df.has_facebook.sum()

159920

In [55]:
df.has_twitter.sum()

109853

In [56]:
df.head()

Name   ABV  AverageRating  \
0                       Närke Kaggen Stormaktsporter   9.5       4.489771   
1                              Westvleteren 12 (XII)  10.2       4.426578   
2                    Schramm’s The Heart of Darkness  14.0       4.423655   
3  B. Nektar Ken Schramm Signature Series - The H...  14.0       4.421873   
4  Cigar City Pilot Series Dragonfruit Passion Fr...   0.0       4.420719   

   RatingCount        BeerStyle   BreweryType          City     State  \
0          557   Imperial Stout  Microbrewery        Örebro       NaN   
1         3332    Abt/Quadrupel  Microbrewery  Westvleteren       NaN   
2           77             Mead       Meadery      Ferndale  Michigan   
3           50             Mead       Meadery      Ferndale  Michigan   
4           46  Berliner Weisse  Microbrewery         Tampa   Florida   

         Country        Twitter                              Facebook  \
0         Sweden            NaN                                   NaN   
1        Belgium            NaN                                   NaN   
2  United States   schrammsmead  https://facebook.com/SchrammsMeadery   
3  United States        bnektar         https://facebook.com/b.nektar   
4  United States  cigarcitybeer    https://facebook.com/cigarcitybeer   

   has_twitter  has_facebook  
0            0             0  
1            0             0  
2            1             1  
3            1             1  
4            1             1

Now that we have the binary columns we can drop the twitter and facebook columns. 

In [57]:
df.drop(['Twitter','Facebook'],axis=1,inplace=True)

In [58]:
df.head()

Name   ABV  AverageRating  \
0                       Närke Kaggen Stormaktsporter   9.5       4.489771   
1                              Westvleteren 12 (XII)  10.2       4.426578   
2                    Schramm’s The Heart of Darkness  14.0       4.423655   
3  B. Nektar Ken Schramm Signature Series - The H...  14.0       4.421873   
4  Cigar City Pilot Series Dragonfruit Passion Fr...   0.0       4.420719   

   RatingCount        BeerStyle   BreweryType          City     State  \
0          557   Imperial Stout  Microbrewery        Örebro       NaN   
1         3332    Abt/Quadrupel  Microbrewery  Westvleteren       NaN   
2           77             Mead       Meadery      Ferndale  Michigan   
3           50             Mead       Meadery      Ferndale  Michigan   
4           46  Berliner Weisse  Microbrewery         Tampa   Florida   

         Country  has_twitter  has_facebook  
0         Sweden            0             0  
1        Belgium            0             0  
2  United States            1             1  
3  United States            1             1  
4  United States            1             1

### Missing Values in the other columns

We will look at the missing values in the City, State, Country, BreweryType and BeerStyle columns.

In [59]:
df.BeerStyle.isnull().sum()

0

In [60]:
df.BreweryType.isnull().sum()

0

In [67]:
df.Country = df.Country.str.strip()
df.Country.isnull().sum()

0

In [63]:
df.State.isnull().sum()

81306

In [64]:
df.City.isnull().sum()

560

There are 81,306 missing values in the state column and 560 missing values from the city column. We are not interested in comparing the beers from different states and cities around the world. We will evaluate the varations across the countries. However, it will be good to compare beers across states and cities within the US. Let us check if there are states and cities missing for beers from the US

In [70]:
# number of missing states for the beers from the US 
df.State = df.State.str.strip()
df[df.Country == 'United States'].State.isnull().sum()

54

In [71]:
# number of missing cities for the beers from US
df.City = df.City.str.strip()
df[df.Country == 'United States'].City.isnull().sum()

8

In [72]:
# No. of beers from the US 
df[df.Country == 'United States'].Country.count()

88228

There are 54 missing states and 8 cities out of the 88,228 observations. This is not very siginifcant and will keep it in the dataset for now. We can fix as needed during the exploratory data analysis as needed.

The dataframe can be saved as a .csv file now for further analysis.

In [73]:
df.to_csv('Allbeers.csv',sep=',',encoding='utf-8')